src/sql/src/pure.rs

```diff
@@ -1648,56 +1745,69 @@ async fn purify_create_table_from_source(
                 PurifiedExportDetails::LoadGenerator { table, output } => (table, output),
                 _ => unreachable!("purified export details must be load generator"),
             };
-
-            let (gen_columns, gen_constraints) = scx.relation_desc_into_table_defs(&desc)?;
+            // We only determine the table description for multi-output load generator sources here,
+            // whereas single-output load generators will have their relation description
+            // determined during statment planning as envelope and format options may affect their
+            // schema.
+            if let Some(desc) = desc {
+                let (gen_columns, gen_constraints) = scx.relation_desc_into_table_defs(&desc)?;
+                *columns = gen_columns;
+                *constraints = gen_constraints;
+            }
             let details = SourceExportStatementDetails::LoadGenerator { output };
-            *columns = gen_columns;
-            *constraints = gen_constraints;
             with_options.push(TableFromSourceOption {
                 name: TableFromSourceOptionName::Details,
                 value: Some(WithOptionValue::Value(Value::String(hex::encode(
                     details.into_proto().encode_to_vec(),
                 )))),
             })
         }
-        PurifiedExportDetails::Kafka { .. } => {
-            sql_bail!("Kafka sources do not yet support CREATE TABLE .. FROM SOURCE")
+        PurifiedExportDetails::Kafka {} => {
+            // NOTE: Kafka tables have their 'schemas' purified into the statement inside the
+            // format field, so we don't specify any columns or constraints to be stored
+            // on the statement here. The RelationDesc will be determined during planning.
+            let details = SourceExportStatementDetails::Kafka {};
+            with_options.push(TableFromSourceOption {
+                name: TableFromSourceOptionName::Details,
+                value: Some(WithOptionValue::Value(Value::String(hex::encode(
+                    details.into_proto().encode_to_vec(),
+                )))),
+            })
         }
     };
 
     Ok(PurifiedStatement::PurifiedCreateTableFromSource { stmt })
 }
 
+enum SourceFormatOptions {
+    Default,
+    Kafka { topic: String },
+}
+
 async fn purify_source_format(
     catalog: &dyn SessionCatalog,
     format: &mut Option<FormatSpecifier<Aug>>,
-    connection: &mut CreateSourceConnection<Aug>,
+    options: &SourceFormatOptions,
```

Comment:
explaining these changes: This refactor to introduce `SourceFormatOptions` was necessary since I needed to re-use this `purify_source_format` method inside the `purify_create_table_from_source` method which doesn't have a reference to the `CreateSourceConnection<Aug>` object. The only thing that was used for was to know if it was a kafka source and if so, what the topic was, to be used for retrieving the appropriate schema-registry details

Reply:
I see. A bit unfortunate that we have this coupling between determining the schema of the format and the kind of source being used. A way out would be to pass the external reference as the schema registry subject name which coincides with the topic name for kafka sources. Not something we have to do in this PR